src/service/storage-impl.js

```diff
@@ -0,0 +1,381 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {assert} from '../asserts';
+import {getService} from '../service';
+import {getSourceOrigin} from '../url';
+import {isExperimentOn} from '../experiments';
+import {log} from '../log';
+import {timer} from '../timer';
+import {viewerFor} from '../viewer';
+
+/** @const */
+const EXPERIMENT = 'amp-storage';
+
+/** @const */
+const TAG = 'Storage';
+
+/** @const */
+const MAX_VALUES_PER_ORIGIN = 8;
+
+
+/**
+ * The storage API. This is an API equivalent to the Web LocalStorage API but
+ * extended to all AMP embedding scenarios.
+ *
+ * The storage is done per source origin. See `get`, `set` and `remove` methods
+ * for more info.
+ *
+ * Requires "amp-storage" experiment.
+ *
+ * @see https://html.spec.whatwg.org/multipage/webstorage.html
+ * @private Visible for testing only.
+ */
+export class Storage {
+
+  /**
+   * @param {!Window} win
+   * @param {!Viewer} viewer
+   * @param {!StorageBindingDef} binding
+   */
+  constructor(win, viewer, binding) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewer;
+
+    /** @private @const {!StorageBindingDef} */
+    this.binding_ = binding;
+
+    /** @const @private {string} */
+    this.origin_ = getSourceOrigin(this.win.location);
+
+    /** @const @private {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.win, EXPERIMENT);
+
+    /** @private @const {!Promise} */
+    this.whenStarted_ = this.isExperimentOn_ ?
+        Promise.resolve() :
+        Promise.reject(`Enable experiment ${EXPERIMENT}`);
+
+    /** @private {?Promise<!Store>} */
+    this.storePromise_ = null;
+  }
+
+  /**
+   * @return {!Storage}
+   * @private
+   */
+  start_() {
+    if (!this.isExperimentOn_) {
+      log.info(TAG, 'Storage experiment is off: ', EXPERIMENT);
+      return this;
+    }
+    this.listenToBroadcasts_();
+    return this;
+  }
+
+  /**
+   * Returns the promise that yields the value of the property for the specified
+   * key.
+   * @param {string} name
+   * @return {!Promise<*>}
+   * @override
+   */
+  get(name) {
+    return this.getStore_().then(store => store.get(name));
+  }
+
+  /**
+   * Saves the value of the specified property. Returns the promise that's
+   * resolved when the operation completes.
+   * @param {string} name
+   * @param {*} value
+   * @return {!Promise}
+   * @override
+   */
+  set(name, value) {
+    assert(typeof value == 'boolean', 'Only boolean values accepted');
```

Comment:
We need to make sure to not provide a "read all keys" API. That could be used to put values into keys :)